### PR TITLE
Add MAGMA implementations of Torch LAPACK functions

### PR DESCRIPTION
**Pull request by colesbury**
This adds MAGMA binding for a number of LAPACK functions. If MAGMA is not installed, the functions simply raise an error.

### CODE REVIEW

**Comment by soumith** [comment]
this looks largely good to go. It has merge conflicts, because I just merged in a PR that refactors the CMakeLists. Look at this PR: https://github.com/torch/cutorch/pull/174

**Comment by szagoruyko** [comment]
I think all cmake MAGMA logic has to go to lib/THC and FindMAGMA has to be in lib/THC/cmake/CMakeLists.txt
btw should there be a doc how to install it?

File: lib/THC/THCTensorMathMagma.cu [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+  magma_spotri_gpu(MagmaUpper, n, input_data, n, &info);
+  if (info > 0)
+    THError("MAGMA potri : A(%d,%d) is 0, A cannot be factorized", info, info);
+  else if (info < 0)
+    THError("MAGMA potri : Argument %d : illegal value", -info);
+
+  const int len = n*n;
+  cudaStream_t stream = THCState_getCurrentStream(state);
+  THCudaTensor_copyUpperSymmetric<<<(len+127/128), 128, 0, stream>>>(input_data, n, len);
```
Comment (dominikgrewe) [inline lib/THC/THCTensorMathMagma.cu]: Make sure you don't launch more than the maximum number of thread blocks. You can query it from `state->deviceProperties[currentDevice].maxGridSize[0]`.

File: lib/THC/THCTensorMathMagma.cu [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+  return ptr;
+}
+
+static void THCudaTensor_copyArray1d(THCState *state, THCudaTensor *self, float *src, int k)
+{
+  long size[1] = { k };
+  long stride[1] = { 1 };
+  THCudaTensor_rawResize(state, self, 1, size, stride);
+  size_t len = k * sizeof(float);
+  THCudaCheck(cudaMemcpy(self->storage->data + self->storageOffset, src, len, cudaMemcpyHostToDevice));
```
Comment (dominikgrewe) [inline lib/THC/THCTensorMathMagma.cu]: Make sure that `self` is actually contiguous. `THCudaTensor_rawResize` doesn't guarantee that.

Same in the function below.

Comment (colesbury) [inline lib/THC/THCTensorMathMagma.cu]: I believe it does ensure that self is contiguous since it ensures the stride is 1.

Comment (dominikgrewe) [inline lib/THC/THCTensorMathMagma.cu]: Fair point. Didn't think about the stride you're passing in.

**Comment by colesbury** [comment]
I've fixed the bug @dominikgrewe pointed out where a large input could result in the block dim overflowing.

I also rebased and THC should still be separately buildable. The FindMAGMA is required by both cutorch and THC since THCGeneral.h.in defines USE_MAGMA based on whether the package is found.

**Comment by soumith** [comment]
Are there any more reservations in merging this? Sam addressed dominik's and sergey's comments.

**Comment by soumith** [comment]
@colesbury merge conflicts again :(

**Comment by colesbury** [comment]
I rebased

**Pull request merged**